### ISSUE TITLE
Have Support for Symlinks in Rclone

### ISSUE THREAD

**Issue by khurshid-alam** [ncw/rclone#40]
`copy` or `sync` command does **NOT** work for symlink(soft links) folders when syncing to Google Drive. It would be nice if rclone can follow symlinks.

**Comment by zeshanb** [comment]
Hello there,

Did you define remote for local file system also?

Regards,
Zeshan

**Comment by khurshid-alam** [comment]
> Did you define remote for local file system also?

What did you mean by that?

The command I used is something like this (on Ubuntu 14.04):
`rclone sync /home/$user/$Test drive:Documents/$Test`. Here the folder `$Test` consists several symlinks to folders on different partition. Rclone exits without following symlinks. Am I missing something?

**Comment by zeshanb** [comment]
Sorry, I mean using rclone config command. There is a command line wizard.

rclone config

Regards,
Zeshan

**Comment by khurshid-alam** [comment]
> Sorry, I mean using rclone config command. There is a command line wizard
rclone config

Yes. That was already done.

**Comment by zeshanb** [comment]
Will have to wait for Nick. Sorry, I wasn't much help.

Regards,
Zeshan

**Comment by ncw** [comment]
@khurshid-alam rclone (taking the lead from rsync) deliberately doesn't follow symlinks as doing so then opens it up to recursive directory loops and other nasty things like that

I could implement this flag (from the rsync man page)

       -L, --copy-links
              When symlinks are encountered, the item that they point to  (the
              referent) is copied, rather than ignoring the symlink

However rclone will never be able to copy symlinks themselves (you can't store these on object storage systems), so it will work like rsync with the `-L` flag and without the `-K` flag.

The `-L` flag would only work on local filesystems.

Eg if your source directory looks like this

```
   root
      link -> /path/to/dir

   /path/to/dir
      file one
      file two
```

The destination would end up looking like this

```
  root
    link
        file one
        file two
```

Does that match your expectations?

Possibly this should be the default behaviour of rclone - I'm not sure.

**Comment by khurshid-alam** [comment]
> However rclone will never be able to copy symlinks themselves (you can't store these on object storage systems), so it will work like rsync with the -L flag and without the -K flag.

>The -L flag would only work on local filesystems.

That's is fine with me. I intend to use it only for Local-->Remote sync only.

>The destination would end up looking like this
```
root
    link
        file one
        file two
```
Does that match your expectations?

Yes it is! I usually rename my symlinks folders to match the actual ones. This is perfect for me.

**Comment by skottie-miller** [comment]
I have a slightly different spin on this request.  I'd like a way to preserve symlinks in POSIX
filesystems when doing a copy/sync to (or from) object store.

in addition to a flag that follows links and copies the content of the linked target,
I'd like to see a flag that copies symlinks as "symlink object".

Since you can't represent a symlink in an object, you instead create a placeholder
object (lets call it an ".slink" object, that contains the metadata needed to re-create
the symlink when its copied out of the object store.

for example,  if a source directory contains
ls -l bazinga 
lrwxrwxrwx 1 skottie user 20 May 26 16:54 bazinga -> /some/symlinked/path

an object would be created in the destination object store called "bazinga.slink",
and inside that object would be the string "/some/symlinked/path" and anything else
needed to re-create the link on the target filesystem.

or something like that.

 -skottie

**Comment by kkniffin** [comment]
I would also be interested in this feature. I have several directories that have symlinks to other locations on external storage. It would be nice to have the rclone follow those links and copy the contents from those directories.

**Comment by wernight** [comment]
+1

Just need to be careful in case of recursive symlink that it gets copied only once.

Test example:

````
temp1/
  a1/
    b1/
       a2/ --> ../../a2/
       f1
  a2/
    b2/
       a1/ --> ../../a1/
       f2
  a3/ --> ../temp2/
temp2/
  f3
````

Expected copied files structure (without cycles):
 * a1/b1/f1
 * a2/b2/f2
 * a3/f3

(note `temp1/a1/b1/a2/b2/a1/b1/a2/b2/a1/b1/a2/b2/` is valid and some directory traversal would happily loop there)

**Comment by wernight** [comment]
See https://github.com/GoogleCloudPlatform/gsutil/pull/325 for how this can be implemented in Python.

**Comment by debackerl** [comment]
Please, be sure to keep both options. I'm actually requiring rclone to skip symlinks. Thank you :-)

If you want to keep symlinks on cloud hosting, maybe you could save the target path as a standard text file, with identical filename, and a special flag in metadatas when the provider supports metadatas.

**Comment by wernight** [comment]
It's not about keeping symlinks as cloud hosts usually don't allow symlinks. It's about mimicking `rsync -k` (and even better by handling circular references) so that all files can get copied (backed up). It's also an optional flag.

**Comment by y2kbug-hk** [comment]
-L or --copy-links is not currently supported yet, isn't it?

**Comment by wernight** [comment]
Correct. In [that PR](https://github.com/GoogleCloudPlatform/gsutil/pull/325) I've actually improved `-L` to detect, alert, and nicely handle circular references (rare but can happen and rsync is just having a kind of path overflow there).

**Comment by darylstyrk** [comment]
I came here to recommend --copy-links. It's a must for my rsync's so hopefully the PR will get some attention.

**Comment by ShapeShifter499** [comment]
Any updates on this?  I'm looking forward to a solution to this.

**Comment by ncw** [comment]
I'll try to get this done for v1.35

**Comment by darylstyrk** [comment]
Has this feature been included yet?

**Comment by ncw** [comment]
@darylstyrk not yet - I haven't released v1.34 yet, this should be in v1.35.  I'll post a beta here in due course for people to try

**Comment by gedl** [comment]
+1 for this feature. It's really handy to be able to create a directory with some links to things I want to backup and then sync the directory holding the links
The alternative is to run separate commands for each link target, which get's out of hand quite quickly

**Comment by markpwoodward** [comment]
+1 for the "--links", where the symlink is somehow copied (e.g. text file with the same name and metadata for the target). It would be a pain when restoring to have to try and recreate all of the symlinks from memory. I sync to google drive.

p.s. thank you for such a great tool!

**Comment by theluke** [comment]
Strongly reuqest!

**Comment by dimbeni** [comment]
We would also love symlink support!

**Comment by datslo** [comment]
This would be a game changer!

**Comment by ncw** [comment]
I have implemented `-L`, `--copy-links` as per rsync.  This effectively follows symlinks.

Please find a beta here

http://beta.rclone.org/v1.35-48-g94947f2/ (uploaded in 15-30 mins)

If anyone is still interested in `-l`, `--links` then can you make a new ticket please.

**Comment by kcwu** [comment]
Thanks. It works.

**Comment by theluke** [comment]
Hi folks, how do I test this ? Do i need to overwrite my current rclone installation?

**Comment by ncw** [comment]
@theluke You can download it to a seperate place and use the full path to invoke rclone, or overwrite your current installation whichever you choose.

**Comment by maakuth** [comment]
I wonder if my issue with the newest beta is connected to this ticket. I'm getting:

``` ./rclone sync --bwlimit=8M /data/src/ b2:dst
2017/02/27 09:53:19 ERROR : B2 bucket dst: not deleting files as there were IO errors
2017/02/27 09:53:19 ERROR : Attempt 1/3 failed with 0 errors and: error reading source directory "subdir": "subdir/1": is a not a regular file
2017/02/27 09:53:26 ERROR : B2 bucket dst: not deleting files as there were IO errors
2017/02/27 09:53:26 ERROR : Attempt 2/3 failed with 0 errors and: error reading source directory "subdir": "subdir/1": is a not a regular file
2017/02/27 09:53:32 ERROR : B2 bucket dst: not deleting files as there were IO errors
2017/02/27 09:53:32 ERROR : Attempt 3/3 failed with 0 errors and: error reading source directory "subdir": "subdir/1": is a not a regular file
2017/02/27 09:53:32 Failed to sync: error reading source directory "subdir": "subdir/1": is a not a regular file
```

`subdir/1`  is a symlink elsewhere. This didn't produce an error before. I'm not interested in enabling symlink traversal, as the links point to another subdirectory inside the source directory. Link traversal would therefore do double copies of the linked directories.

**Comment by ncw** [comment]
@maakuth you are right - there is a small mistake in a recent commit - thanks for reporting.

Try this - this should fix it

http://beta.rclone.org/v1.35-149-g5fba913/ (uploaded in 15-30 mins)

**Comment by maakuth** [comment]
> Try this - this should fix it

Seems to work, thank you!

**Comment by tjcinnamon** [comment]
@ncw I'm on version 1.35 and I keep getting an unknown shorthand flag 'L'

` rclone copy '/my Dir/File 1.pptx' drive:remoteFolder -L`

Do I have the correct version?  Is my command copying symlinks correct?

**Comment by tjcinnamon** [comment]
@ncw Thanks for the response!  The roadmap is for it to end up in 1.36, correct?

**Comment by ncw** [comment]
@tjcinnamon yes, which should (cross fingers) be out this weekend.

**Comment by philou669** [comment]
Hi,

I'm on rclone-v1.57.0-linux-amd64. I'm not very good in English and I don't understand where I have to put -L or --copy-links in my command line to avoid this "Can't follow symlink without -L/--copy-links"


this is my current command
rclone copy --create-empty-src-dirs --log-file=/mnt/md0/public/auto/log/docker_config.txt /mnt/md0/appdata/docker/config/ crypt:/docker_config

help me please
thanks

**Comment by ncw** [comment]
@philou669 please [post on the rclone forum](https://forum.rclone.org/) about this - thank you
